### ISSUE TITLE
torch with cloned networks has a problem during updating parameters

### ISSUE THREAD

**Issue by yjy765** [torch/torch7#716]
It is really similar to old issues that soumith raised a long time ago.

I made a network that contains some shared elements as following

    mlp1 = nn.Sequential()
    mlp1:add(nn.Linear(3,2))
    mlp2 = mlp1:clone('weight','bias','gradWeight','gradBias')
    
    model = nn.ConcatTable()
    model:add(mlp1)
    model:add(mlp2)

then, I made some artificial data composed of input and gradOutput

Forward and backward function work well and gradients of parameters are correctly estimated.

But, it has a problem when I call updateParameters()

When I call model:updateParameters(lr) then the parameters are subtracted by twice of calculated gradient. If I added three cloned network then three times of gradients are subtracted.

Can anybody explain and solve it?

**Comment by iamalbert** [comment]
just leave the gradParams (`gradWeight` and `gradBias`) unshared.

**Comment by yjy765** [comment]
@iamalbert  thank you for your reply.
but I already tried to leave gradParam unshared. but it causes an error when calling getParameters()
the getParameters() in Module.lua make me ensure the element of parameters and gradParameters should be equal.
The above example exactly do not need to call getParameters() but to use optim package or something else I should call getParameters().
Also to check the validity of updateParameters() I should call getParameters().
Is it okay even though I cannot call getParameters()?

**Comment by szagoruyko** [comment]
@yjy765 I don't see a problem here, what do you expect to happen?

**Comment by yjy765** [comment]
@szagoruyko 
ok say, the intial value of the parameters is [-0.1,-0.1,-0.1,-0.1,-0.1,-0.1,-0.1,-0.1] and backward function results in gradParam to be [9,1,1,1,1,1,1,1].
Then when I call updateParameters(1), I expect parameters to be [-9.1,-1.1,-1.1,-1.1,-1.1,-1.1,-1.1,-1.1,-1.1].
However, it has two linear layers and the parameter become [-18.1,-2.2,-2.2,-2.2,-2.2,-2.2,-2.2].

If the number of sharing is small, it can be ignored, but I use lots of shared layers.

**Comment by timharley** [comment]
Use the gradParams vector returned by getParameters() to do a step of SGD,
rather than module:updateParameters(). This is what torch/optim does.

**Comment by yjy765** [comment]
@timharley ok, thank you for reply again
You mean make a opfunc and use optim package?
As far as I understand, I tried to use optim package as follows

    opfunc = function(x)
        if x~=param then
           param:copy(x)
        end
        model:forward(some input)
        model:backward(some input, some gradOutput)
        return some error, gradParam
    end
    optim.sgd(opfunc,param,config)

However, it shows same results as I used getParameters()

I may not understand what you mean anyway

**Comment by timharley** [comment]
@yjy765 Something like this, if you wanted to implement SGD manually:

```
model = ...
params, gradParams = model:getParameters()

-- Body of training loop:
gradParams:zero()
model:forward(...)
model:backward(...)
params:add(-lr, gradParams)
```

`model:updateParameters()` assumes that none of the parameters are shared.

Using `torch/optim` instead of manually doing SGD allows more powerful optimisation.

**Comment by yjy765** [comment]
@timharley thank u so much
Yes, it works. I made a mistake in the above code with optim.sgd
You are correct. updateParameters does not consider sharing so I should use torch/optim or
follow as you did above.
Great thanks!!

**Comment by Amir-Arsalan** [comment]
@yjy765 @soumith I have been trying to figure out how to resolve the parameter sharing problem in my network architecture. I'm not using SGD and have to use one of the optim packages' methods (ADAM). I've posted a simplified version of my code [here](http://stackoverflow.com/questions/38725285/unable-to-do-parameter-sharing-in-torch-between-networks). Can you guys check my code out and hopefully give me some hints on how I should go about solving the issue?

**Comment by Amir-Arsalan** [comment]
@yjy765 @soumith  I have been trying to figure out how to resolve the parameter sharing problem in my network architecture. I'm not using SGD and have to use one of the optim packages' methods (ADAM). I've posted a simplified version of my code [here](http://stackoverflow.com/questions/38725285/unable-to-do-parameter-sharing-in-torch-between-networks/38727530#38727530). Can you guys check my code out and hopefully give me some hints on how I should go about solving the issue?

**Comment by soumith** [comment]
@Amir-Arsalan https://github.com/torch/torch7/issues/716#issuecomment-237049877
